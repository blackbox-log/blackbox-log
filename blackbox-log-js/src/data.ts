import { Memoize as memoize } from 'typescript-memoize';

import { WasmPointer } from './wasm';

import type { Headers } from './headers';
import type { WasmExports, WasmObject } from './wasm';

export class DataParser implements WasmObject {
	readonly #wasm: WasmExports;
	readonly #ptr: WasmPointer;
	readonly #headers: Headers;

	constructor(wasm: WasmExports, headers: Headers, ptr: number) {
		this.#wasm = wasm;
		this.#headers = headers;
		this.#ptr = new WasmPointer(ptr, wasm.data_free);
	}

	free() {
		this.#ptr.free();
	}

	get isAlive(): boolean {
		return this.#ptr.isAlive;
	}

	get headers(): Headers {
		return this.#headers;
	}

	parseAll() {
		this.#wasm.data_parseAll(this.#ptr.ptr);
	}

	get mainFrameCount(): number {
		return this.#wasm.data_mainFrameCount(this.#ptr.ptr);
	}

	get gpsFrameCount(): number {
		return this.#wasm.data_gpsFrameCount(this.#ptr.ptr);
	}
}
