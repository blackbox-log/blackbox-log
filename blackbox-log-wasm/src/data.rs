use blackbox_log::prelude::*;
use blackbox_log::Reader;

use crate::{OwnedSlice, Shared, WasmFfi};

// SAFETY: field order *must* be `parser` first, then `headers`, then `data` to
// ensure correct drop order
pub struct WasmDataParser {
    parser: DataParser<'static, 'static>,
    _headers: Shared<Headers<'static>>,
    _data: Shared<OwnedSlice>,
}

impl WasmDataParser {
    pub(crate) fn new(
        headers: Shared<Headers<'static>>,
        reader: Reader<'static>,
        data: Shared<OwnedSlice>,
    ) -> Self {
        // SAFETY: this is only used to create the `DataParser`, which is guaranteed to
        // be dropped before `headers` by the declaration order in the struct
        let headers_ref = unsafe { headers.deref_static() };

        Self {
            parser: DataParser::new(reader, headers_ref),
            _headers: headers,
            _data: data,
        }
    }

    fn main_frame_count(&self) -> usize {
        self.parser.stats().counts.main
    }

    fn gps_frame_count(&self) -> usize {
        self.parser.stats().counts.gps
    }
}

impl WasmFfi for WasmDataParser {}

#[no_mangle]
unsafe extern "wasm" fn data_free(ptr: *mut WasmDataParser) {
    let parser = WasmDataParser::from_wasm(ptr);
    drop(parser);
}

#[no_mangle]
#[allow(non_snake_case)]
unsafe extern "wasm" fn data_mainFrameCount(ptr: *mut WasmDataParser) -> usize {
    let parser = WasmDataParser::from_wasm(ptr);
    let count = parser.main_frame_count();
    parser.into_wasm();
    count
}

#[no_mangle]
#[allow(non_snake_case)]
unsafe extern "wasm" fn data_gpsFrameCount(ptr: *mut WasmDataParser) -> usize {
    let parser = WasmDataParser::from_wasm(ptr);
    let count = parser.gps_frame_count();
    parser.into_wasm();
    count
}
