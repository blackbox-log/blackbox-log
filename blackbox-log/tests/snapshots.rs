use std::collections::BTreeMap;
use std::fs;
use std::io::Read;

use blackbox_log::data::{ParseEvent, Stats};
use blackbox_log::event::Event;
use blackbox_log::units::{si, Flag, FlagSet};
use blackbox_log::{DataParser, Headers, Unit, Value};
use serde::ser::{SerializeStruct, Serializer};
use serde::Serialize;

macro_rules! run {
    () => {
        |path| {
            let mut file = fs::File::open(path).unwrap();
            let mut data = Vec::new();
            file.read_to_end(&mut data).unwrap();

            let file = blackbox_log::File::new(&data);
            let logs = file
                .iter()
                .map(|mut reader| {
                    Headers::parse(&mut reader).map(|headers| {
                        let data = DataParser::new(&mut reader, &headers);
                        LogSnapshot::new(&headers, data)
                    })
                })
                .collect::<Vec<_>>();

            insta::assert_ron_snapshot!(logs);
        }
    };
}

#[test]
fn own() {
    insta::glob!("logs/*.bbl", run!());
}

#[test]
#[ignore]
fn fc_blackbox() {
    insta::glob!("logs/fc-blackbox/*", run!());
}

#[test]
#[ignore]
fn gimbal_ghost() {
    insta::glob!("logs/gimbal-ghost/*", run!());
}

#[derive(Debug, Serialize)]
struct LogSnapshot<'data> {
    headers: Headers<'data>,
    stats: Stats,
    events: Vec<Event>,
    main: Fields,
    slow: Fields,
    gps: Fields,
}

impl<'data> LogSnapshot<'data> {
    fn new(headers: &Headers<'data>, mut data: DataParser<'data, '_, '_>) -> Self {
        let headers = headers.clone();

        let mut events = Vec::new();
        let mut main = headers.main_def().iter().collect::<Fields>();
        let mut slow = headers.slow_def().iter().collect::<Fields>();
        let mut gps = headers
            .gps_def()
            .iter()
            .flat_map(|def| def.iter())
            .collect::<Fields>();

        while let Some(frame) = data.next() {
            match frame {
                ParseEvent::Event(event) => events.push(event),
                ParseEvent::Main(frame) => main.update(frame),
                ParseEvent::Slow(frame) => slow.update(frame),
                ParseEvent::Gps(frame) => gps.update(frame),
            }
        }

        Self {
            headers,
            stats: data.stats().clone(),
            events,
            main,
            slow,
            gps,
        }
    }
}

#[derive(Debug, Serialize)]
struct Fields(Vec<FieldSnapshot>);

impl Fields {
    fn update<F: blackbox_log::frame::Frame>(&mut self, frame: F) {
        for (field, value) in self.0.iter_mut().zip(frame.iter()) {
            field.update(value.into());
        }
    }
}

impl<T, U> FromIterator<(T, U)> for Fields
where
    T: Into<String>,
    U: Into<Unit>,
{
    fn from_iter<I: IntoIterator<Item = (T, U)>>(iter: I) -> Self {
        Self(
            iter.into_iter()
                .map(|(name, unit)| FieldSnapshot::new(name.into(), unit.into()))
                .collect(),
        )
    }
}

#[derive(Debug, Serialize)]
struct FieldSnapshot {
    name: String,
    unit: Unit,
    history: History,
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum History {
    Int(NumberHistory<16, i128>),
    Bool { yes: usize, no: usize },
    Flags(BTreeMap<&'static str, usize>),
}

#[derive(Debug)]
struct NumberHistory<const N: usize, T> {
    min: T,
    max: T,
    seen: Vec<T>,
    histogram: [usize; N],
}

impl FieldSnapshot {
    fn new(name: String, unit: Unit) -> Self {
        Self {
            name,
            unit,
            history: match unit {
                Unit::FrameTime
                | Unit::Amperage
                | Unit::Voltage
                | Unit::Acceleration
                | Unit::Rotation
                | Unit::GpsCoordinate
                | Unit::Altitude
                | Unit::Velocity
                | Unit::GpsHeading
                | Unit::Unitless => History::Int(NumberHistory::new()),
                Unit::Boolean => History::Bool { yes: 0, no: 0 },
                Unit::FlightMode | Unit::State | Unit::FailsafePhase => {
                    History::Flags(BTreeMap::new())
                }
            },
        }
    }

    #[allow(clippy::wildcard_enum_match_arm, clippy::cast_possible_truncation)]
    fn update(&mut self, value: Value) {
        match &mut self.history {
            History::Int(history) => history.update(match value {
                Value::FrameTime(t) => t.get::<si::time::microsecond>().round() as i128,
                Value::Amperage(a) => a.get::<si::electric_current::milliampere>().round() as i128,
                Value::Voltage(v) => v.get::<si::electric_potential::millivolt>().round() as i128,
                Value::Acceleration(a) => a
                    .get::<si::acceleration::centimeter_per_second_squared>()
                    .round() as i128,
                Value::Rotation(r) => {
                    r.get::<si::angular_velocity::degree_per_second>().round() as i128
                }
                Value::GpsCoordinate(c) => (c * 10000000.).round() as i128,
                Value::Altitude(a) => a.get::<si::length::meter>().round() as i128,
                Value::Velocity(v) => {
                    v.get::<si::velocity::centimeter_per_second>().round() as i128
                }
                Value::GpsHeading(h) => (h * 10.).round() as i128,
                Value::Unsigned(u) => u.into(),
                Value::Signed(s) => s.into(),
                _ => unreachable!(),
            }),
            History::Bool { yes, no } => match value {
                Value::Boolean(true) => *yes += 1,
                Value::Boolean(false) => *no += 1,
                _ => unreachable!(),
            },
            History::Flags(history) => {
                if let Value::FailsafePhase(phase) = value {
                    *history.entry(phase.as_name()).or_insert(0) += 1;
                } else {
                    let flags = match value {
                        Value::FlightMode(m) => m.as_names(),
                        Value::State(s) => s.as_names(),
                        _ => unreachable!(),
                    };

                    for flag in flags {
                        *history.entry(flag).or_insert(0) += 1;
                    }
                }
            }
        }
    }
}

impl<const N: usize, T> NumberHistory<N, T> {
    fn new() -> Self
    where
        T: Default,
    {
        Self {
            min: T::default(),
            max: T::default(),
            seen: Vec::new(),
            histogram: [0; N],
        }
    }

    fn update_range(&mut self, value: T)
    where
        T: PartialOrd,
    {
        if value < self.min {
            self.min = value;
        } else if value > self.max {
            self.max = value;
        }
    }

    /// Store new value sorted in `seen` if not already present
    fn update_seen(&mut self, value: T)
    where
        T: PartialOrd,
    {
        let index = self.seen.partition_point(|x| *x < value);
        if self.seen.get(index) != Some(&value) {
            self.seen.insert(index, value);
        }
    }

    /// Bucket so that sequential values go in sequential buckets, since data is
    /// usually clustered
    #[inline(always)]
    fn update_histogram(&mut self, value: u128) {
        self.histogram[(value % N as u128) as usize] += 1;
    }
}

impl<const N: usize> NumberHistory<N, i128> {
    fn update(&mut self, value: i128) {
        self.update_range(value);
        self.update_seen(value);
        self.update_histogram(value.unsigned_abs());
    }
}

impl<const N: usize> Serialize for NumberHistory<N, i128>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(self, self.min, self.max, serializer)
    }
}

impl<const N: usize> Serialize for NumberHistory<N, f64>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(
            self,
            format!("{:.2}", self.min),
            format!("{:.2}", self.max),
            serializer,
        )
    }
}

#[inline(always)]
fn serialize<const N: usize, T, S, U>(
    snapshot: &NumberHistory<N, T>,
    min: U,
    max: U,
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: Serializer,
    U: Serialize,
    [usize; N]: Serialize,
{
    let unique = snapshot.seen.len();

    let mut state = serializer.serialize_struct("FieldSnapshot", 4)?;

    state.serialize_field("min", &min)?;
    state.serialize_field("max", &max)?;
    state.serialize_field("unique", &unique)?;

    let histogram = "histogram";
    if unique > 1 {
        state.serialize_field(histogram, &snapshot.histogram)?;
    } else {
        state.skip_field(histogram)?;
    }

    state.end()
}
