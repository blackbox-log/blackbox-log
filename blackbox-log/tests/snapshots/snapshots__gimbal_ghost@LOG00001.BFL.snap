---
source: blackbox-log/tests/snapshots.rs
expression: logs
input_file: ".modules/gimbal-ghost/test/LOG00001.BFL"
---
[
  Ok(LogSnapshot(
    headers: Headers(
      version: V2,
      firmware_revision: "Betaflight 4.3.0 (fc8318f54) STM32F405",
      firmware_kind: Betaflight,
      board_info: Some("MTKS MATEKF405CTR"),
      craft_name: None,
      vbat_reference: Some(1629),
      vbat_scale: Some(110),
      current_meter: Some(CurrentMeterConfig(
        offset: 0,
        scale: 179,
      )),
      acceleration_1g: Some(2048),
      gyro_scale: Some(0.017453292),
      min_throttle: Some(1070),
      motor_output_range: Some(MotorOutputRange(158, 2047)),
      unknown: {
        "acc_hardware": "0",
        "dterm_notch_cutoff": "0",
        "pid_process_denom": "1",
        "gyro_cal_on_first_arm": "0",
        "yaw_lowpass_hz": "100",
        "acc_limit_yaw": "0",
        "dyn_notch_max_hz": "600",
        "yaw_deadband": "0",
        "mag_hardware": "1",
        "simplified_dterm_filter_multiplier": "100",
        "dshot_bidir": "1",
        "dterm_lpf1_dyn_hz": "75,150",
        "simplified_feedforward_gain": "100",
        "gyro_sync_denom": "1",
        "simplified_dterm_filter": "1",
        "serialrx_provider": "12",
        "throttle_limit_type": "0",
        "dshot_idle_value": "550",
        "dyn_idle_p_gain": "50",
        "rc_smoothing_debug_axis": "0",
        "motor_output_limit": "100",
        "gyro_to_use": "0",
        "dterm_lpf2_type": "0",
        "vbatcellvoltage": "330,350,430",
        "tpa_breakpoint": "1350",
        "rpm_filter_harmonics": "3",
        "throttle_limit_percent": "100",
        "rates": "67,67,67",
        "acc_limit": "0",
        "d_min": "30,34,0",
        "feedforward_averaging": "0",
        "gyro_lpf1_static_hz": "250",
        "d_max_gain": "37",
        "dyn_idle_min_rpm": "0",
        "gyro_lpf1_dyn_hz": "250,500",
        "motor_poles": "14",
        "acc_lpf_hz": "1000",
        "vbat_sag_compensation": "0",
        "baro_hardware": "0",
        "rc_rates": "7,7,7",
        "pid_at_min_throttle": "1",
        "d_max_advance": "20",
        "P interval": "8",
        "rates_type": "3",
        "dyn_idle_i_gain": "50",
        "pidsum_limit_yaw": "400",
        "pidsum_limit": "500",
        "dyn_notch_min_hz": "150",
        "motor_pwm_rate": "480",
        "rc_smoothing_auto_factor_throttle": "30",
        "dyn_notch_q": "500",
        "dterm_notch_hz": "0",
        "rollPID": "45,80,40",
        "rc_smoothing_setpoint_cutoff": "0",
        "yawPID": "45,80,0",
        "levelPID": "50,50,75",
        "airmode_activate_throttle": "25",
        "rc_smoothing_feedforward_cutoff": "0",
        "gyro_hardware_lpf": "0",
        "anti_gravity_gain": "3500",
        "rc_smoothing_rx_average": "9059",
        "dyn_idle_max_increase": "150",
        "feedforward_max_rate_limit": "90",
        "rc_smoothing_auto_factor": "30",
        "iterm_windup": "85",
        "gyro_notch_cutoff": "0,0",
        "throttle_boost_cutoff": "15",
        "simplified_pitch_d_gain": "100",
        "rc_smoothing_active_cutoffs_ff_sp_thr": "41,41,41",
        "deadband": "0",
        "tpa_rate": "65",
        "rate_limits": "1998,1998,1998",
        "simplified_d_gain": "100",
        "rpm_filter_fade_range_hz": "50",
        "maxthrottle": "2000",
        "feedforward_smooth_factor": "25",
        "I interval": "256",
        "use_integrated_yaw": "0",
        "simplified_dmax_gain": "100",
        "simplified_gyro_filter": "1",
        "gyro_lpf1_type": "0",
        "features": "272893128",
        "fields_disabled_mask": "0",
        "rpm_filter_min_hz": "100",
        "dyn_idle_d_gain": "50",
        "simplified_pids_mode": "2",
        "Log start datetime": "2022-02-01T16:16:32.235+00:00",
        "gyro_lpf2_type": "0",
        "anti_gravity_mode": "0",
        "simplified_master_multiplier": "100",
        "iterm_relax": "1",
        "dyn_notch_count": "1",
        "gyro_notch_hz": "0,0",
        "iterm_relax_cutoff": "15",
        "simplified_pitch_pi_gain": "100",
        "rc_smoothing_throttle_cutoff": "0",
        "feedforward_boost": "15",
        "dterm_lpf1_type": "0",
        "motor_pwm_protocol": "7",
        "gyro_lpf2_static_hz": "500",
        "anti_gravity_threshold": "250",
        "rpm_filter_q": "500",
        "rc_smoothing": "1",
        "dterm_lpf1_static_hz": "75",
        "pitchPID": "47,84,46",
        "simplified_pi_gain": "100",
        "magPID": "40",
        "simplified_gyro_filter_multiplier": "100",
        "feedforward_jitter_factor": "7",
        "ff_weight": "120,125,120",
        "Firmware date": "Jan 30 2022 01:43:14",
        "iterm_relax_type": "1",
        "simplified_i_gain": "100",
        "rpm_filter_lpf_hz": "150",
        "thr_mid": "50",
        "throttle_boost": "5",
        "feedforward_transition": "0",
        "thr_expo": "0",
        "looptime": "125",
        "rc_expo": "0,0,0",
        "debug_mode": "6",
        "P ratio": "32",
        "dterm_lpf2_static_hz": "150",
        "use_unsynced_pwm": "0",
        "abs_control_gain": "0",
      },
    ),
    stats: Stats(
      counts: FrameCounts(
        main: 229265,
        slow: 35,
        gps: 1170,
        gps_home: 0,
      ),
    ),
    events: [
      SyncBeep(385573225),
      FlightMode(
        flags: 0,
        last_flags: 1,
      ),
      Disarm(4),
      End,
    ],
    main: Fields([
      FieldSnapshot(
        name: "loopIteration",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 1834000,
          unique: 229265,
          histogram: (14330, 14329, 14329, 14329, 14329, 14329, 14329, 14329, 14329, 14329, 14329, 14329, 14329, 14329, 14329, 14329),
        ),
      ),
      FieldSnapshot(
        name: "time",
        unit: FrameTime,
        history: FieldSnapshot(
          min: 0,
          max: 621006493,
          unique: 229265,
          histogram: (14147, 14275, 14262, 14331, 14350, 14343, 14229, 14441, 14291, 14402, 14125, 14412, 14529, 14362, 14292, 14474),
        ),
      ),
      FieldSnapshot(
        name: "axisP[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -116,
          max: 121,
          unique: 174,
          histogram: (49755, 43263, 33570, 25301, 18649, 14235, 10931, 8493, 6478, 4920, 3950, 3027, 2368, 1804, 1386, 1135),
        ),
      ),
      FieldSnapshot(
        name: "axisP[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -72,
          max: 57,
          unique: 119,
          histogram: (52282, 44052, 33729, 25292, 18677, 13699, 10295, 7988, 5982, 4605, 3526, 2735, 2163, 1767, 1373, 1100),
        ),
      ),
      FieldSnapshot(
        name: "axisP[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -63,
          max: 41,
          unique: 84,
          histogram: (112439, 59783, 27486, 12752, 6452, 3750, 2035, 1261, 878, 723, 560, 400, 259, 207, 158, 122),
        ),
      ),
      FieldSnapshot(
        name: "axisI[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -62,
          max: 69,
          unique: 132,
          histogram: (28842, 24122, 21912, 22127, 21367, 18828, 17989, 14415, 11558, 9195, 8051, 7756, 6631, 6321, 5279, 4872),
        ),
      ),
      FieldSnapshot(
        name: "axisI[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -67,
          max: 111,
          unique: 179,
          histogram: (19679, 16236, 15418, 14866, 14002, 12802, 13758, 15625, 16047, 14656, 15250, 14716, 14019, 11885, 10841, 9465),
        ),
      ),
      FieldSnapshot(
        name: "axisI[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -84,
          max: 113,
          unique: 198,
          histogram: (24683, 20202, 18546, 17901, 17833, 17474, 15705, 14797, 13142, 13194, 10300, 10488, 10238, 9349, 7686, 7727),
        ),
      ),
      FieldSnapshot(
        name: "axisD[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -115,
          max: 146,
          unique: 194,
          histogram: (33568, 31011, 27771, 24140, 20402, 17091, 14172, 12146, 10320, 8475, 7162, 6233, 5380, 4390, 3784, 3220),
        ),
      ),
      FieldSnapshot(
        name: "axisD[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -112,
          max: 74,
          unique: 132,
          histogram: (42054, 37825, 31437, 25433, 20106, 15872, 12551, 9982, 8038, 6391, 5221, 4195, 3329, 2744, 2276, 1811),
        ),
      ),
      FieldSnapshot(
        name: "axisF[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -61,
          max: 38,
          unique: 99,
          histogram: (187540, 15871, 10115, 5599, 3263, 2109, 1460, 875, 644, 496, 380, 350, 248, 124, 103, 88),
        ),
      ),
      FieldSnapshot(
        name: "axisF[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -23,
          max: 25,
          unique: 49,
          histogram: (199374, 12511, 8068, 3590, 2070, 1448, 722, 473, 235, 231, 127, 115, 70, 87, 65, 79),
        ),
      ),
      FieldSnapshot(
        name: "axisF[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -49,
          max: 69,
          unique: 115,
          histogram: (205073, 9481, 5888, 3275, 1607, 1261, 761, 441, 353, 289, 211, 190, 176, 116, 86, 57),
        ),
      ),
      FieldSnapshot(
        name: "rcCommand[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -300,
          max: 245,
          unique: 365,
          histogram: (31173, 15303, 25171, 10150, 16847, 9128, 15921, 7040, 19181, 7172, 17915, 9170, 15450, 6901, 15357, 7386),
        ),
      ),
      FieldSnapshot(
        name: "rcCommand[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -159,
          max: 129,
          unique: 244,
          histogram: (44856, 17721, 26065, 9550, 20791, 8287, 15809, 9808, 15137, 4437, 14434, 4946, 14002, 5145, 13719, 4558),
        ),
      ),
      FieldSnapshot(
        name: "rcCommand[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -142,
          max: 329,
          unique: 322,
          histogram: (26108, 24321, 28428, 8527, 22978, 8335, 14878, 7238, 16961, 8544, 15400, 5807, 16080, 6722, 12118, 6820),
        ),
      ),
      FieldSnapshot(
        name: "rcCommand[3]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 1844,
          unique: 837,
          histogram: (11807, 15053, 21247, 11921, 14855, 15862, 9801, 16087, 16640, 14131, 20029, 11317, 10644, 16782, 14169, 8920),
        ),
      ),
      FieldSnapshot(
        name: "setpoint[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -258,
          max: 178,
          unique: 422,
          histogram: (35894, 17118, 17170, 19794, 15575, 12876, 12412, 11316, 10050, 11997, 14610, 11850, 10453, 7933, 9609, 10608),
        ),
      ),
      FieldSnapshot(
        name: "setpoint[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -83,
          max: 58,
          unique: 142,
          histogram: (55035, 26036, 16131, 20537, 15968, 12482, 9833, 9453, 8425, 9622, 7213, 9235, 9139, 6687, 6627, 6842),
        ),
      ),
      FieldSnapshot(
        name: "setpoint[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -69,
          max: 305,
          unique: 355,
          histogram: (48607, 26006, 19463, 14151, 12676, 10010, 10612, 13378, 12197, 8331, 9327, 9255, 11073, 8732, 7985, 7462),
        ),
      ),
      FieldSnapshot(
        name: "setpoint[3]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 848,
          unique: 840,
          histogram: (16641, 13581, 20291, 11959, 10727, 16255, 13858, 9347, 11771, 15452, 20958, 11862, 14591, 15819, 9854, 16299),
        ),
      ),
      FieldSnapshot(
        name: "vbatLatest",
        unit: Voltage,
        history: FieldSnapshot(
          min: 0,
          max: 17520,
          unique: 444,
          histogram: (27154, 0, 27759, 0, 29663, 0, 28149, 0, 29271, 0, 27895, 0, 30536, 0, 28838, 0),
        ),
      ),
      FieldSnapshot(
        name: "amperageLatest",
        unit: Amperage,
        history: FieldSnapshot(
          min: 0,
          max: 75580,
          unique: 782,
          histogram: (29228, 0, 29338, 0, 30926, 0, 27502, 0, 28291, 0, 28149, 0, 27798, 0, 28033, 0),
        ),
      ),
      FieldSnapshot(
        name: "BaroAlt",
        unit: Unitless,
        history: FieldSnapshot(
          min: -381,
          max: 9556,
          unique: 5303,
          histogram: (14441, 15067, 13533, 13988, 15025, 14249, 14629, 14493, 13263, 15360, 14918, 14316, 13601, 15432, 12716, 14234),
        ),
      ),
      FieldSnapshot(
        name: "rssi",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 918,
          unique: 521,
          histogram: (16376, 12728, 16851, 14176, 13554, 16229, 11896, 17667, 11234, 16883, 12768, 13922, 14875, 11838, 16478, 11790),
        ),
      ),
      FieldSnapshot(
        name: "gyroADC[0]",
        unit: Rotation,
        history: FieldSnapshot(
          min: -270,
          max: 188,
          unique: 448,
          histogram: (18769, 26347, 21388, 19041, 16961, 15278, 13482, 12464, 11763, 11700, 11702, 11462, 10606, 9839, 9402, 9061),
        ),
      ),
      FieldSnapshot(
        name: "gyroADC[1]",
        unit: Rotation,
        history: FieldSnapshot(
          min: -103,
          max: 68,
          unique: 169,
          histogram: (25569, 37049, 27380, 21105, 17167, 14575, 12249, 11373, 10308, 9647, 9121, 7960, 7419, 6702, 6115, 5526),
        ),
      ),
      FieldSnapshot(
        name: "gyroADC[2]",
        unit: Rotation,
        history: FieldSnapshot(
          min: -71,
          max: 309,
          unique: 369,
          histogram: (30581, 38486, 24702, 16853, 12083, 10993, 10533, 10982, 10659, 9968, 10045, 10329, 10152, 8850, 7071, 6978),
        ),
      ),
      FieldSnapshot(
        name: "accSmooth[0]",
        unit: Acceleration,
        history: FieldSnapshot(
          min: -1554,
          max: 445,
          unique: 1833,
          histogram: (14492, 14832, 15258, 14572, 14770, 14224, 14232, 15031, 14089, 13862, 13754, 14457, 13833, 14120, 13879, 13860),
        ),
      ),
      FieldSnapshot(
        name: "accSmooth[1]",
        unit: Acceleration,
        history: FieldSnapshot(
          min: -1045,
          max: 1362,
          unique: 2158,
          histogram: (14337, 14691, 15134, 14571, 14652, 14404, 14607, 15046, 14472, 14386, 13785, 14676, 13801, 13770, 13623, 13310),
        ),
      ),
      FieldSnapshot(
        name: "accSmooth[2]",
        unit: Acceleration,
        history: FieldSnapshot(
          min: -298,
          max: 4576,
          unique: 4305,
          histogram: (14285, 14393, 14446, 14376, 14249, 14334, 14337, 14166, 14275, 14464, 14512, 14340, 14200, 14363, 14302, 14223),
        ),
      ),
      FieldSnapshot(
        name: "debug[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -471,
          max: 332,
          unique: 643,
          histogram: (15161, 17934, 17132, 15952, 15624, 15377, 14803, 14263, 14059, 13778, 13561, 13051, 12550, 12426, 11963, 11631),
        ),
      ),
      FieldSnapshot(
        name: "debug[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -173,
          max: 103,
          unique: 248,
          histogram: (15316, 20719, 19014, 17843, 17298, 16476, 15355, 14667, 14040, 13374, 12505, 12039, 11224, 10332, 9850, 9213),
        ),
      ),
      FieldSnapshot(
        name: "debug[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -124,
          max: 318,
          unique: 434,
          histogram: (15410, 20848, 18983, 17673, 17250, 16134, 15399, 14502, 13606, 12981, 12374, 11673, 11339, 10878, 10356, 9859),
        ),
      ),
      FieldSnapshot(
        name: "debug[3]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 0,
          unique: 1,
        ),
      ),
      FieldSnapshot(
        name: "motor[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 2057,
          unique: 1877,
          histogram: (14168, 14369, 14299, 14321, 14328, 13969, 14294, 14483, 14209, 14155, 14152, 14326, 14427, 14521, 14819, 14425),
        ),
      ),
      FieldSnapshot(
        name: "motor[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 1955,
          unique: 1602,
          histogram: (14450, 14124, 14203, 14216, 14519, 14350, 14608, 13989, 14150, 14215, 14159, 14526, 14496, 14353, 14516, 14391),
        ),
      ),
      FieldSnapshot(
        name: "motor[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 1986,
          unique: 1770,
          histogram: (14258, 14003, 14248, 14203, 14248, 14442, 14189, 14281, 14267, 14368, 14323, 14232, 14513, 14700, 14441, 14549),
        ),
      ),
      FieldSnapshot(
        name: "motor[3]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 1956,
          unique: 1662,
          histogram: (14408, 14260, 14282, 14328, 14331, 14277, 14152, 14267, 14354, 14290, 14402, 14581, 14316, 14297, 14487, 14233),
        ),
      ),
      FieldSnapshot(
        name: "flightModeFlags",
        unit: FlightMode,
        history: {
          "Arm": 229228,
        },
      ),
      FieldSnapshot(
        name: "stateFlags",
        unit: State,
        history: {
          "GpsFix": 221994,
          "GpsFixEver": 229264,
        },
      ),
      FieldSnapshot(
        name: "failsafePhase",
        unit: FailsafePhase,
        history: {
          "Idle": 229264,
        },
      ),
      FieldSnapshot(
        name: "rxSignalReceived",
        unit: Boolean,
        history: History(
          yes: 229264,
          no: 0,
        ),
      ),
      FieldSnapshot(
        name: "rxFlightChannelsValid",
        unit: Boolean,
        history: History(
          yes: 229264,
          no: 0,
        ),
      ),
    ]),
    gps: Fields([
      FieldSnapshot(
        name: "time",
        unit: FrameTime,
        history: FieldSnapshot(
          min: 0,
          max: 620991448,
          unique: 1169,
          histogram: (75, 69, 74, 56, 78, 85, 69, 64, 79, 61, 81, 75, 65, 76, 77, 85),
        ),
      ),
      FieldSnapshot(
        name: "GPS_numSat",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 6,
          unique: 4,
          histogram: (37, 0, 0, 0, 481, 351, 300, 0, 0, 0, 0, 0, 0, 0, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "GPS_coord[0]",
        unit: GpsCoordinate,
        history: FieldSnapshot(
          min: -337981507,
          max: 0,
          unique: 1084,
          histogram: (76, 69, 70, 77, 86, 90, 72, 81, 67, 79, 78, 52, 73, 63, 69, 67),
        ),
      ),
      FieldSnapshot(
        name: "GPS_coord[1]",
        unit: GpsCoordinate,
        history: FieldSnapshot(
          min: 0,
          max: 201019858,
          unique: 1102,
          histogram: (76, 66, 85, 71, 71, 68, 69, 88, 75, 82, 71, 61, 54, 83, 77, 72),
        ),
      ),
      FieldSnapshot(
        name: "GPS_altitude",
        unit: Altitude,
        history: FieldSnapshot(
          min: 0,
          max: 3747,
          unique: 717,
          histogram: (72, 66, 73, 78, 63, 70, 69, 70, 86, 63, 66, 76, 70, 65, 94, 88),
        ),
      ),
      FieldSnapshot(
        name: "GPS_speed",
        unit: Velocity,
        history: FieldSnapshot(
          min: 0,
          max: 2806,
          unique: 893,
          histogram: (86, 94, 85, 59, 69, 62, 80, 80, 77, 74, 74, 77, 63, 61, 63, 65),
        ),
      ),
      FieldSnapshot(
        name: "GPS_ground_course",
        unit: GpsHeading,
        history: FieldSnapshot(
          min: 0,
          max: 3598,
          unique: 789,
          histogram: (70, 60, 106, 80, 100, 50, 86, 55, 80, 61, 85, 44, 61, 101, 63, 67),
        ),
      ),
    ]),
  )),
]
